pymc3/backends/report.py

```diff
@@ -0,0 +1,145 @@
+from collections import namedtuple
+import logging
+
+
+logger = logging.getLogger('pymc3')
+
+
+SamplerWarning = namedtuple(
+    'SamplerWarning',
+    "kind, message, level, step, exec_info, extra")
+
+
+_LEVELS = {
+    'info': logging.INFO,
+    'error': logging.ERROR,
+    'warn': logging.WARN,
+    'debug': logging.DEBUG,
+    'critical': logging.CRITICAL,
+}
+
+
+class SamplerReport(object):
+    def __init__(self):
+        self._chain_warnings = {}
+        self._global_warnings = []
+        self._effective_n = None
+        self._gelman_rubin = None
+
+    @property
+    def _warnings(self):
+        chains = sum(self._chain_warnings.values(), [])
+        return chains + self._global_warnings
+
+    @property
+    def ok(self):
+        """Whether the automatic convergence checks found serious problems."""
+        return all(_LEVELS[warn.level] < _LEVELS['warn']
+                   for warn in self._warnings)
+
+    def raise_ok(self, level='error'):
+        errors = [warn for warn in self._warnings
+                  if _LEVELS[warn.level] >= _LEVELS[level]]
+        if errors:
+            raise ValueError('Serious convergence issues during sampling.')
+
+    def _run_convergence_checks(self, trace):
+        if trace.nchains == 1:
+            msg = ("Only one chain was sampled, this makes it impossible to "
+                   "run some convergence checks")
+            warn = SamplerWarning('bad-params', msg, 'info', None, None, None)
+            self._add_warnings([warn])
+            return
+
+        from pymc3 import diagnostics
+
+        self._effective_n = effective_n = diagnostics.effective_n(trace)
+        self._gelman_rubin = gelman_rubin = diagnostics.gelman_rubin(trace)
+
+        warnings = []
+        rhat_max = max(val.max() for val in gelman_rubin.values())
+        if rhat_max > 1.4:
+            msg = ("The gelman-rubin statistic is larger than 1.4 for some "
+                   "parameters. The sampler did not converge.")
+            warn = SamplerWarning(
+                'convergence', msg, 'error', None, None, gelman_rubin)
+            warnings.append(warn)
+        elif rhat_max > 1.2:
+            msg = ("The gelman-rubin statistic is larger than 1.2 for some "
+                   "parameters.")
+            warn = SamplerWarning(
+                'convergence', msg, 'warn', None, None, gelman_rubin)
+            warnings.append(warn)
+        elif rhat_max > 1.05:
+            msg = ("The gelman-rubin statistic is larger than 1.05 for some "
+                   "parameters. This indicates slight problems during "
+                   "sampling.")
+            warn = SamplerWarning(
+                'convergence', msg, 'info', None, None, gelman_rubin)
+            warnings.append(warn)
+
+        eff_min = min(val.min() for val in effective_n.values())
```

Comment:
would it be more informative to raise warning if the effect sample size is lower than a certain percentage of the total sample? For example, raise warning if effect sample size is below 25% of the total number of sample.

Reply:
Good idea. We could have both, too. One with level `info` if it is <25% and one with level `warn` if it is <200?

Reply:
I would do `warn` if it is <25% (high autocorrelation and poor mixing is likely an indication of modelling problem) and `info` if it is <200 (sometimes ppl sample only 200 for demo purpose and it would be annoying to see warning)

Reply:
I guess that comes down to how much we trust `effective_n`. If we don't trust the numbers it returns, but only use it as an indication that the sampler doesn't work well, then I think I agree. But if we take the values from `effective_n` at face value, then a low percentage of effective samples isn't in itself a problem (ie should only be `info`). But a very low number of eff samples is a problem.
Say your samples have some autocorrelation, but you don't get any divergences, and gelman_rubin looks fine. Then I don't see a problem with just running the sampler for a very long time, until you get a lot of effective samples. This seems to me like a valid use-case, and we'd print warnings if people did this, even though what they are doing is ok. (at least I think it is, right?)
On the other hand, few effective samples are always a problem if you plan to do anything with your trace.
Maybe a middle way would be:
- `info` warning if `neff < 0.25 * draws`
- `warn` if `neff < 200` *and* `draws > 500`

That way we don't issue a low-neff warning if users *explicitly* asked for a low number of samples.

Reply:
Yeah that's a good point, I agree with this solution.